titus-server-master/src/main/java/com/netflix/titus/master/kubernetes/controller/NodeGcController.java

```diff
@@ -0,0 +1,180 @@
+                .filter(this::isNodeEligibleForGc)
+                .map(n -> KubeUtil.getMetadataName(n.getMetadata()))
+                .filter(StringExt::isNotEmpty)
+                .collect(Collectors.toSet());
+
+        logger.info("Attempting to GC {} nodes: {}", nodeNamesToGc.size(), nodeNamesToGc);
+        this.nodesToGc.set(nodeNamesToGc.size());
+        for (String nodeName : nodeNamesToGc) {
+            gcNode(nodeName);
+        }
+        logger.info("Finished node GC");
+    }
+
+    @VisibleForTesting
+    boolean isNodeEligibleForGc(V1Node node) {
+        String nodeName = KubeUtil.getMetadataName(node.getMetadata());
+        Optional<V1NodeCondition> readyNodeConditionOpt = KubeUtil.findNodeCondition(node, READY);
+        if (StringExt.isNotEmpty(nodeName) && readyNodeConditionOpt.isPresent()) {
+            V1NodeCondition readyNodeCondition = readyNodeConditionOpt.get();
+            boolean isReadyConditionTimestampPastGracePeriod = hasConditionGracePeriodElapsed(readyNodeCondition, configuration.getNodeGcGracePeriodMs());
+            boolean isAgentInstanceNotAvailable = isAgentInstanceNotAvailable(nodeName);
+            return isReadyConditionTimestampPastGracePeriod && isAgentInstanceNotAvailable;
+        }
+        return false;
+    }
+
+    private boolean hasConditionGracePeriodElapsed(V1NodeCondition condition, long gracePeriodMs) {
+        DateTime lastHeartbeatTime = condition.getLastHeartbeatTime();
+        return lastHeartbeatTime != null && clock.isPast(lastHeartbeatTime.getMillis() + gracePeriodMs);
+    }
+
+    private boolean isAgentInstanceNotAvailable(String nodeName) {
+        if (StringExt.isEmpty(nodeName)) {
+            return false;
+        }
+
+        Optional<AgentInstance> agentInstanceOpt = agentManagementService.findAgentInstance(nodeName);
+        if (!agentInstanceOpt.isPresent()) {
+            return true;
+        }
+
+        AgentInstance agentInstance = agentInstanceOpt.get();
+        return agentInstance.getLifecycleStatus().getState() == InstanceLifecycleState.Stopped;
+    }
+
+    private void gcNode(String nodeName) {
+        if (StringExt.isNotEmpty(nodeName)) {
+            try {
+                kubeApiFacade.getCoreV1Api().deleteNode(
+                        nodeName,
+                        null,
+                        null,
+                        0,
+                        null,
+                        BACKGROUND,
+                        null
+                );
+            } catch (JsonSyntaxException e) {
+                // this is probably successful. the generated client has the wrong response type
+            } catch (ApiException e) {
+                if (!e.getMessage().equalsIgnoreCase(NOT_FOUND)) {
+                    logger.error("Failed to delete node: {} with error: ", nodeName, e);
+                }
+            } catch (Exception e) {
+                logger.error("Failed to delete node: {} with error: ", nodeName, e);
```

Comment:
`CodeInvariants#unexpectedError()`?

Reply:
I wouldn't say this is unexpected as there could be networking/client errors that happen.